generator/collector_camera.cpp

```diff
@@ -0,0 +1,130 @@
+#include "generator/collector_camera.hpp"
+
+#include "generator/feature_builder.hpp"
+#include "generator/osm_element.hpp"
+#include "generator/maxspeeds_parser.hpp"
+
+#include "routing/routing_helpers.hpp"
+
+#include "routing_common/maxspeed_conversion.hpp"
+
+#include "indexer/ftypes_matcher.hpp"
+
+#include "platform/measurement_utils.hpp"
+
+#include "coding/point_coding.hpp"
+#include "coding/write_to_sink.hpp"
+
+#include "geometry/latlon.hpp"
+
+#include "base/assert.hpp"
+#include "base/logging.hpp"
+#include "base/string_utils.hpp"
+
+namespace routing
+{
+size_t const CameraProcessor::kMaxSpeedSpeedStringLength = 32;
+
+std::string ValidateMaxSpeedString(std::string const & maxSpeedString)
+{
+  routing::SpeedInUnits speed;
+  if (!generator::ParseMaxspeedTag(maxSpeedString, speed) || !speed.IsNumeric())
+    return std::string();
+
+  return strings::to_string(measurement_utils::ToSpeedKmPH(speed.GetSpeed(), speed.GetUnits()));
+}
+
+CameraProcessor::CameraInfo::CameraInfo(OsmElement const & element)
+  : m_id(element.id)
+  , m_lon(element.lon)
+  , m_lat(element.lat)
+{
+  auto const maxspeed = element.GetTag("maxspeed");
+  if (!maxspeed.empty())
+    m_speed = ValidateMaxSpeedString(maxspeed);
+}
+
+void CameraProcessor::ForEachCamera(Fn && toDo) const
+{
+  std::vector<uint64_t> empty;
+  for (auto const & p : m_speedCameras)
+  {
+    auto const & ways = m_cameraToWays.count(p.first) != 0 ? m_cameraToWays.at(p.first) : empty;
+    toDo(p.second, ways);
+  }
+}
+
+void CameraProcessor::ProcessWay(OsmElement const & element)
+{
+  for (auto const node : element.m_nds)
+  {
+    if (m_speedCameras.find(node) == m_speedCameras.cend())
```

Comment:
то что всё это будет работать обеспечивается тем, что в формате o5m сначала идут все node, потом way
для osm.xml это не так, но мы уже и в некоторых других местах osm.xml не поддерживаем (например информацию про road access собираем аналогично)
предлагаю написать где-то коммент об этом (что поддерживается только o5m) и в целом задуматься не выпилить ли нам поддержку xml

Reply:
согласен, тут все завязано на порядок
предлагаю позже добавить гарантию порядка в xml
глядя на формат в вики это кажется возможным

Reply:
ну и коммент не повредит: вряд ли всем так уж очевидно какие гарантии у o5m на порядок следования элементов, а без этого знания не понятно почему вообще всё будет работать. Типа как в RoadAccess:

``` C++
271 void RoadAccessTagProcessor::Process(OsmElement const & elem, ofstream & oss)
272 {
273   // We will process all nodes before ways because of o5m format:
274   // all nodes are first, then all ways, then all relations.
```

Reply:
> согласен, тут все завязано на порядок
> предлагаю позже добавить гарантию порядка в xml
> глядя на формат в вики это кажется возможным

мб реально стоит забить на xml лучше пользоваться какой-нибудь оберткой из osmconvert, чтобы делать что-то вроде
`Reader GetO5MReaderFromXML(str::string const & XMLContent)`

Reply:
или так, посмотрим как будет быстрее и проще

Reply:
++ оставил коммент